lib/model/model.go

```diff
@@ -1016,23 +1016,18 @@ func (m *model) handleIndex(deviceID protocol.DeviceID, folder string, fs []prot
 	}
 
 	m.fmut.RLock()
-	files, existing := m.folderFiles[folder]
-	runner, running := m.folderRunners[folder]
+	files, ok := m.folderFiles[folder]
+	runner := m.folderRunners[folder]
 	m.fmut.RUnlock()
 
-	if !existing {
-		l.Warnf("%v for nonexistent folder %q", op, folder)
-		panic("handling index for nonexistent folder")
+	if !ok {
+		// This can happen due to a config change that isn't yet
+		// reflected in the model
+		l.Debugf("%v for folder %q missing in model", op, folder)
+		return
```

Comment:
Here's a question though... What are those circumstances exactly? Can it be that we're not up and running yet so we miss an index update and then process later ones, thus permanently lack index info? Or is this just when we're tearing things down, or what?

Reply:
Race on folder restart (e.g. unpause): After removing the folder from model and closing the connection, you need to acquire fmut to readd the folder in which time the connection can already be established and index messages incoming. However looking at this again I think we should be able to do all of it (removing, adding and starting) while continuously holding fmut - thus preventing this race.

The race on adding a folder however I believe we can't prevent (easily): Folder is added in cfg, some unrelated change closes the connection, which then gets readded before the folder is added in the model. The only way I see to prevent this is to newer use the wrapper directly in the model (specifically in `generateClusterConfig`), but copy everything to model during `CommitConfiguration`. Thus we'd always have a consistent state within model. Here it means even if the connection comes in before we add the folder, we don't send the info about the new folder in the CC, thus won't receive index updates preemptively.